File: src/tools/registry.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { GraphQLClient } from 'graphql-request';
import { z } from 'zod';
import { REGISTRY_LIST, REGISTRY_CREATE, REGISTRY_DELETE } from '../queries/registry.js';

export function registerRegistryTools(server: McpServer, client: GraphQLClient): void {
  server.registerTool(
    'nexaa_registry_list',
    {
      description: 'List all private registry connections in a namespace',
      inputSchema: { namespace: z.string() },
    },
    async ({ namespace }) => {
      const data = await client.request<{
        namespace: { privateRegistries: unknown[] };
      }>(REGISTRY_LIST, { namespaceName: namespace });
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.namespace.privateRegistries, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_registry_create',
    {
      description:
        'Add a private container registry connection. Note: password is write-only and will not be returned in responses.',
      inputSchema: {
        namespace: z.string(),
        name: z.string(),
        source: z.string().describe('Registry URL e.g. registry.example.com'),
        username: z.string(),
        password: z.string(),
        verify: z.boolean().default(true).describe('Verify registry credentials on creation'),
      },
    },
    async (input) => {
      const data = await client.request<{ registryConnectionCreate: unknown }>(REGISTRY_CREATE, {
        input,
      });
      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(data.registryConnectionCreate, null, 2),
          },
        ],
      };
    },
  );

  server.registerTool(
    'nexaa_registry_delete',
    {
      description: 'Remove a private registry connection',
      inputSchema: {
        namespace: z.string(),
        name: z.string(),
      },
    },
    async ({ namespace, name }) => {
      await client.request(REGISTRY_DELETE, {
        namespaceName: namespace,
        registryName: name,
      });
      return {
        content: [
          {
            type: 'text',
            text: `Registry "${name}" deleted from namespace "${namespace}".`,
          },
        ],
      };
    },
  );
}

File: tools/registry.go
package tools

import (
	"context"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nexaa-cloud/nexaa-cli/api"
)

func RegisterRegistry(s *server.MCPServer) {
	s.AddTool(
		mcp.NewTool("nexaa_registry_list",
			mcp.WithDescription("List all private registry connections in a namespace"),
			mcp.WithString("namespace", mcp.Required()),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			ns := req.GetString("namespace", "")
			result, err := withAPI(func(c *api.Client) ([]api.RegistryResult, error) {
				return c.ListRegistries(ns)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_registry_create",
			mcp.WithDescription("Add a private container registry connection. Note: password is write-only and will not be returned in responses."),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("name", mcp.Required()),
			mcp.WithString("source", mcp.Required(), mcp.Description("Registry URL e.g. registry.example.com")),
			mcp.WithString("username", mcp.Required()),
			mcp.WithString("password", mcp.Required()),
			mcp.WithBoolean("verify", mcp.Description("Verify registry credentials on creation"), mcp.DefaultBool(true)),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			var args api.RegistryCreateInput
			if err := req.BindArguments(&args); err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			// mcp.DefaultBool only annotates the JSON schema; BindArguments is a
			// plain json.Unmarshal that doesn't apply it, so an omitted "verify"
			// would otherwise leave args.Verify at its Go zero value (false)
			// instead of the documented default of true.
			if _, ok := req.GetArguments()["verify"]; !ok {
				args.Verify = true
			}
			result, err := withAPI(func(c *api.Client) (api.RegistryResult, error) {
				return c.RegistryCreate(args)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return jsonResult(result), nil
		},
	)

	s.AddTool(
		mcp.NewTool("nexaa_registry_delete",
			mcp.WithDescription("Remove a private registry connection. WARNING: This is irreversible. Requires confirm: true."),
			mcp.WithString("namespace", mcp.Required()),
			mcp.WithString("name", mcp.Required()),
			mcp.WithBoolean("confirm", mcp.Required(), mcp.Description("Must be true to confirm deletion")),
		),
		func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
			ns := req.GetString("namespace", "")
			name := req.GetString("name", "")
			if ok, cancelled := confirmDelete(req, "registry connection"); !ok {
				return cancelled, nil
			}
			_, err := withAPI(func(c *api.Client) (bool, error) {
				return c.RegistryDelete(ns, name)
			})
			if err != nil {
				return mcp.NewToolResultError(err.Error()), nil
			}
			return mcp.NewToolResultText(fmt.Sprintf("Registry %q deleted from namespace %q.", name, ns)), nil
		},
	)
}
